vm_phys.c

```diff
@@ -0,0 +1,241 @@
+
+  for (int i = 0; i < page_array_size; i++) {
+    seg->page_array[i].phys_addr = seg->start + PAGESIZE * i;
+    seg->page_array[i].order = 0;
+    seg->page_array[i].flags = 0;
+  }
+
+  for (int i = 0; i < VM_NFREEORDER; i++)
+    TAILQ_INIT(PG_FREEQ(seg, i));
+  int last_page = 0;
+  int seg_size = page_array_size;
+  for (int ord = VM_NFREEORDER - 1; ord >= 0; ord--)
+    while (POW2(ord) <= seg_size) {
+      vm_page_t *page = &seg->page_array[last_page];
+      page->flags |= VM_FREE;
+      page->order = ord;
+      TAILQ_INSERT_HEAD(PG_FREEQ(seg, page->order), page, freeq);
+      seg_size -= POW2(ord);
+      last_page += POW2(ord);
+    }
+}
+
+void vm_phys_init() {
+  TAILQ_INIT(&seglist);
+}
+
+/* Takes two pages which are buddies, and merges them */
+static vm_page_t *merge(vm_page_t *page1, vm_page_t *page2) {
+  if (page1->order != page2->order)
+    panic("trying to connect pages of different orders");
+  if (page1 > page2) {
+    vm_page_t *tmp = page1;
+    page1 = page2;
+    page2 = tmp;
+  }
+  if (page1 + POW2(page1->order) != page2)
+    panic("pages are not buddies");
+  page1->order++;
+  return page1;
+}
+
+static vm_page_t *find_buddy(vm_phys_seg_t *seg, vm_page_t *page) {
+  vm_page_t *buddy = NULL;
+
+  /* Page is left buddy if it's addres is divisible
+   * by 2^(page->order+1), otherwise it's right buddy */
+
+  if ((page - seg->page_array) % POW2(page->order + 1) == 0)
+    buddy = page + POW2(page->order);
+  else
+    buddy = page - POW2(page->order);
+
+  if (buddy->order != page->order)
+    return NULL;
+
+  /* Check if buddy is on freelist */
+  vm_page_t *it;
+
+  TAILQ_FOREACH(it, PG_FREEQ(seg, page->order), freeq)
+    if (it == buddy)
+      break;
+
+  return buddy;
+}
+
+static void split_page(vm_phys_seg_t *seg, vm_page_t *page) {
+  int order = page->order;
+
+  if (TAILQ_EMPTY(PG_FREEQ(seg, order)))
+    panic("cannot split segment");
+
+  /* Works because page is some indice of page_array */
+  vm_page_t *buddy = &page[1 << (order - 1)];
+
+  TAILQ_REMOVE(PG_FREEQ(seg, page->order), page, freeq);
+
+  page->order = order - 1;
+  buddy->order = order - 1;
+
+  TAILQ_INSERT_HEAD(PG_FREEQ(seg, page->order), page, freeq);
+  TAILQ_INSERT_HEAD(PG_FREEQ(seg, buddy->order), buddy, freeq);
+  buddy->flags |= VM_FREE;
+}
+
+static void vm_phys_reserve_from_seg(vm_phys_seg_t *seg, vm_paddr_t start,
+                                     vm_paddr_t end) {
+  for (int i = VM_NFREEORDER - 1; i >= 0; i--) {
+    vm_page_t *pg_ptr;
+    vm_page_t *pg_it = TAILQ_FIRST(PG_FREEQ(seg, i));
+    while (pg_it) {
+      if (PG_START(pg_it) >= start && PG_END(pg_it) <= end) {
+        /* if segment is containted within (start, end) remove it from free
+         * queue */
+        pg_ptr = pg_it;
+        while (pg_ptr++ < pg_it + POW2(pg_it->order))
+          pg_ptr->flags |= VM_RESERVED;
+        TAILQ_REMOVE(PG_FREEQ(seg, pg_it->order), pg_it, freeq);
+        /* Go to start of list because it's corrupted */
```

Comment:
You should not worry.

Usualy when you iterate over some structure, and modify it while traversing it, the iterator isn't reliable anymore. So the safest way to keep on traversing is to simply start doing it from the very start. And this is what i am doing here. So actually this comment is quite wrong, because list is totally ok, it's more like iterator is corrupted, or more like 'the way of iterating'.  In my opinion i should be able to remove iterator while iterating, because removing out of list doesn't change pointers of iterator, but pointers of elements adjacent to iterator, but i didn't want to rely on this, and (i don't know why) it didn't work.

In my opinion, the best idea would be to create set of elements to be deleted, but this means i'd need to use some kind of dynamic allocation to keep necessary information. 

So basically, i chose the safest solution.